Sources/SourceKit/SourceKitServer.swift

```diff
@@ -338,6 +340,47 @@ extension SourceKitServer {
     toolchainTextDocumentRequest(req, workspace: workspace, fallback: nil)
   }
 
+  func findWorkspaceSymbols(matching: String) -> [SymbolOccurrence] {
+    var symbolOccurenceResults: [SymbolOccurrence] = []
+    workspace?.index?.forEachCanonicalSymbolOccurrence(
+      containing: matching,
+      anchorStart: false,
+      anchorEnd: false,
+      subsequence: true,
+      ignoreCase: true
+    ) {symbol in
+      if !symbol.location.isSystem && !symbol.roles.contains(.accessorOf) {
+        symbolOccurenceResults.append(symbol)
+      }
+      return true
+    }
+    return symbolOccurenceResults
+  }
+
+  func workspaceSymbols(_ req: Request<WorkspaceSymbolsRequest>, workspace: Workspace) {
+    let symbols = findWorkspaceSymbols(
+      matching: req.params.query
+    ).map({symbolOccurrence -> SymbolInformation in
+      let symbolPosition = Position(
+        line: symbolOccurrence.location.line - 1, // 1-based -> 0-based
+        // FIXME: we need to convert the utf8/utf16 column, which may require reading the file!
+        utf16index: symbolOccurrence.location.utf8Column - 1)
+
+      let symbolLocation = Location(
+        url: URL(fileURLWithPath: symbolOccurrence.location.path),
+        range: Range(symbolPosition))
+      // TODO: symbol kind, detail, and containerName should be added
```

Comment:
I would like some guidance on how to get kind, detail, and containerName.

Reply:
* kind: We have a symbol kind in the index that we should expose on `class Symbol`. Then we can convert it to an LSP `SymbolKind` here.

* detail: This isn't a field on `SymbolInformation` AFAICT.  Did you mix it up with `DocumentSymbol`?

* containerName:  Ah didn't realize we'd need this.  For this we would need to expose the `SymbolOccurrence`'s relations in the index.  The C++ version of this is `SymbolOccurrence::getRelations`, we just need to thread it through the C and Swift interfaces.  Since you already have the canonical occurrence, you would find the related symbol that has relation role `childOf` and use that symbol's name.

Reply:
You are correct that detail came from `DocumentSymbol`.

for containerName, it looks like I will have to pass an ArrayPointer to Swift, which I can't find a good example of. Maybe I could use `unsafeBufferPointer` as mentioned [here](https://stackoverflow.com/questions/39666832/converting-c-array-ptr-to-native-array-in-swift-3), but it looks like that involves copying which I think we want to avoid. is there a better way to do this?

Reply:
Copying should be fine, since the number of relations for a single occurrence will generally be small, and we can do the copy lazily only when you ask for related symbols.  Here's a sketch of the C API I would start with.  We can always optimize this later if needed.

```c++
typedef void *indexstoredb_symbol_relation_t;

...

INDEXSTOREDB_PUBLIC uint64_t
indexstoredb_symbol_relation_get_roles(_Nonnull  indexstoredb_symbol_relation_t);

INDEXSTOREDB_PUBLIC indexstoredb_symbol_t
indexstoredb_symbol_relation_get_symbol(_Nonnull indexstoredb_symbol_relation_t);

/// The relations are owned by the occurrence and shall not be used after the occurrence is freed.
INDEXSTOREDB_PUBLIC bool
indexstoredb_symbol_occurrence_relations(_Nonnull indexstoredb_occurrence_t,
                      bool(^applier)(indexstoredb_symbol_relation_t));
```

On the Swift side, this can lazily create an array from the occurrence.